File: src/main/java/com/balatro/Lock.java
package com.balatro;

import com.balatro.api.Named;

import java.util.HashSet;
import java.util.Set;

sealed class Lock permits Functions {

    private final Set<String> locked = new HashSet<>();

    public void lock(String item) {
        locked.add(item);
    }

    public void unlock(String item) {
        locked.remove(item);
    }

    public void lock(Named named) {
        lock(named.getName());
    }

    public boolean isLocked(Named named) {
        return isLocked(named.getName());
    }

    public boolean isLocked(String item) {
        return locked.contains(item);
    }

    public void initLocks(int ante, boolean freshProfile, boolean freshRun) {
        if (ante < 2) {
            lock("The Mouth");
            lock("The Fish");
            lock("The Wall");
            lock("The House");
            lock("The Mark");
            lock("The Wheel");
            lock("The Arm");
            lock("The Water");
            lock("The Needle");
            lock("The Flint");
            lock("Negative Tag");
            lock("Standard Tag");
            lock("Meteor Tag");
            lock("Buffoon Tag");
            lock("Handy Tag");
            lock("Garbage Tag");
            lock("Ethereal Tag");
            lock("Top-up Tag");
            lock("Orbital Tag");
        }

        if (ante < 3) {
            lock("The Tooth");
            lock("The Eye");
        }

        if (ante < 4) lock("The Plant");
        if (ante < 5) lock("The Serpent");
        if (ante < 6) lock("The Ox");

        if (freshProfile) {
            lock("Negative Tag");
            lock("Foil Tag");
            lock("Holographic Tag");
            lock("Polychrome Tag");
            lock("Rare Tag");
            lock("Golden Ticket");
            lock("Mr. Bones");
            lock("Acrobat");
            lock("Sock and Buskin");
            lock("Swashbuckler");
            lock("Troubadour");
            lock("Certificate");
            lock("Smeared Joker");
            lock("Throwback");
            lock("Hanging Chad");
            lock("Rough Gem");
            lock("Bloodstone");
            lock("Arrowhead");
            lock("Onyx Agate");
            lock("Glass Joker");
            lock("Showman");
            lock("Flower Pot");
            lock("Blueprint");
            lock("Wee Joker");
            lock("Merry Andy");
            lock("Oops! All 6s");
            lock("The Idol");
            lock("Seeing Double");
            lock("Matador");
            lock("Hit the Road");
            lock("The Duo");
            lock("The Trio");
            lock("The Family");
            lock("The Order");
            lock("The Tribe");
            lock("Stuntman");
            lock("Invisible Joker");
            lock("Brainstorm");
            lock("Satellite");
            lock("Shoot the Moon");
            lock("Driver's License");
            lock("Cartomancer");
            lock("Astronomer");
            lock("Burnt Joker");
            lock("Bootstraps");
            lock("Overstock Plus");
            lock("Liquidation");
            lock("Glow Up");
            lock("Reroll Glut");
            lock("Omen Globe");
            lock("Observatory");
            lock("Nacho Tong");
            lock("Recyclomancy");
            lock("Tarot Tycoon");
            lock("Planet Tycoon");
            lock("Money Tree");
            lock("Antimatter");
            lock("Illusion");
            lock("Petroglyph");
            lock("Retcon");
            lock("Palette");
        }
        if (freshRun) {
            lock("Planet X");
            lock("Ceres");
            lock("Eris");
            lock("Five of a Kind");
            lock("Flush House");
            lock("Flush Five");
            lock("Stone Joker");
            lock("Steel Joker");
            lock("Glass Joker");
            lock("Golden Ticket");
            lock("Lucky Cat");
            lock("Cavendish");
            lock("Overstock Plus");
            lock("Liquidation");
            lock("Glow Up");
            lock("Reroll Glut");
            lock("Omen Globe");
            lock("Observatory");
            lock("Nacho Tong");
            lock("Recyclomancy");
            lock("Tarot Tycoon");
            lock("Planet Tycoon");
            lock("Money Tree");
            lock("Antimatter");
            lock("Illusion");
            lock("Petroglyph");
            lock("Retcon");
            lock("Palette");
        }
    }

    public void initUnlocks(int ante, boolean freshProfile) {
        if (ante == 2) {
            unlock("The Mouth");
            unlock("The Fish");
            unlock("The Wall");
            unlock("The House");
            unlock("The Mark");
            unlock("The Wheel");
            unlock("The Arm");
            unlock("The Water");
            unlock("The Needle");
            unlock("The Flint");
            if (!freshProfile) unlock("Negative Tag");
            unlock("Standard Tag");
            unlock("Meteor Tag");
            unlock("Buffoon Tag");
            unlock("Handy Tag");
            unlock("Garbage Tag");
            unlock("Ethereal Tag");
            unlock("Top-up Tag");
            unlock("Orbital Tag");
        }
        if (ante == 3) {
            unlock("The Tooth");
            unlock("The Eye");
        }
        if (ante == 4) unlock("The Plant");
        if (ante == 5) unlock("The Serpent");
        if (ante == 6) unlock("The Ox");
    }

}

File: docs/js/immolate2/src/utils/Lock.js
// import { ItemImpl } from "./interface/Item.js";
// import { JokerImpl } from "./interface/Joker.js";
import LOCKED_ITEMS from "../constants/lockedItems";

export class Lock {
    constructor() {
        this.locked = new Set();
    }

    static firstLock = new Set(LOCKED_ITEMS.FIRST_LOCK);

    static ante2Lock = new Set(LOCKED_ITEMS.ANTE_2_LOCK);

    // Helper method to extract common functionality
    processItem(item, action) {
        if (typeof item === 'string') {
            action(item);
        } else if (Array.isArray(item)) {
            item.forEach(i => action(i));
            //TODO implement this for ItemImpl and JokerImpl
            // } else if (item instanceof ItemImpl || item instanceof JokerImpl) {
            //     action(item.getName());
        } else {
            throw new Error("Invalid argument type");
        }
    }

    lock(item) {
        this.processItem(item, name => this.locked.add(name));
    }

    firstLock() {
        Lock.firstLock.forEach(item => this.locked.add(item));;
    }

    unlock(collection) {
       this.processItem(collection, name => this.locked.delete(name));
    }

    isLocked(item) {
        if (typeof item === 'string') {
            return this.locked.has(item);
            // } else if (item instanceof ItemImpl || item instanceof JokerImpl) {
            //     return this.locked.has(item.getName());
        } else {
            throw new Error(`Invalid argument type: ${typeof item}. Expected string, array, or Item/Joker instance.`);
        }
    }

    initLocks(ante, freshProfile, freshRun) {
        if (ante < 2) this.lock(Array.from(Lock.ante2Lock));
        if (ante < 3) this.lock(["The Toot", "The Eye"]);
        if (ante < 4) this.lock("The Plant");
        if (ante < 5) this.lock("The Serpent");
        if (ante < 6) this.lock("The Ox");

        if (freshProfile) {
            this.lock([
                ...LOCKED_ITEMS.FRESH_PROFILE.FIRST_SET,
                ...LOCKED_ITEMS.FRESH_PROFILE.SECOND_SET,
                ...LOCKED_ITEMS.FRESH_PROFILE.THIRD_SET
            ]);
        }

        if (freshRun) {
            this.lock(LOCKED_ITEMS.FRESH_RUN);
        }
    }

    initUnlocks(ante, freshProfile) {
        if (ante === 2) {
            this.unlock([
                "The Mouth", "The Fish", "The Wall", "The House", "The Mark",
                "The Wheel", "The Arm", "The Water", "The Needle", "The Flint"
            ]);
            this.unlock([
                "Standard Tag", "Meteor Tag", "Buffoon Tag", "Handy Tag",
                "Garbage Tag", "Ethereal Tag", "Top-up Tag", "Orbital Tag"
            ]);
            if (!freshProfile) this.unlock("Negative Tag");
        }

        if (ante == 3) {
            this.unlock("The Tooth");
            this.unlock("The Eye");
        }
        if (ante == 4) this.unlock("The Plant");
        if (ante == 5) this.unlock("The Serpent");
        if (ante == 6) this.unlock("The Ox");
    }
}